docs/design/storage_abstraction.md

```diff
@@ -0,0 +1,275 @@
+
+### ObjectStorage
+
+ObjectStorage provides interfaces related to global storage of Objects. It is implemented based on RawTable and is defined as follows:
+
+```move
+module moveos_std::object_storage{
+    #[private_generics(T)]
+    /// Borrow Object from object store with object_id
+    public fun borrow<T: key>(this: &ObjectStorage, object_id: ObjectID): &Object<T>;
+
+    #[private_generics(T)]
+    /// Borrow mut Object from object store with object_id
+    public fun borrow_mut<T: key>(this: &mut ObjectStorage, object_id: ObjectID): &mut Object<T>;
+
+    #[private_generics(T)]
+    /// Remove object from object store
+    public fun remove<T: key>(this: &mut ObjectStorage, object_id: ObjectID): Object<T>;
+
+    #[private_generics(T)]
+    /// Add object to object store
+    public fun add<T: key>(this: &mut ObjectStorage, obj: Object<T>);
+
+    public fun contains<T: key>(this: &mut ObjectStorage, object_id: ObjectID): bool;
+}
+```
+
+The above methods are ensured to only be called directly by Modules containing `T` through [private_generics](https://github.com/rooch-network/rooch/issues/64), and the safety model follow to the constraints of the Move global storage instructions.
+
+>TBD: Whether to provide a method that allows Owners to directly operate on Objects, similar to Sui's `public_transfer`, needs further study.
+
+### TypeTable
+
+TypeTable is a special type of Table that simulates Move's global storage instructions, with the type as the key and an instance of the type as the value. Implemented based on RawTable, it provides the following API:
+
+```move
+module moveos_std::type_table {
+
+    #[private_generics(T)]
+    public fun add<V: key>(table: &mut TypeTable, val: V);
+
+    #[private_generics(T)]
+    public fun borrow<V: key>(table: &TypeTable): &V;
+
+    #[private_generics(T)]
+    public fun borrow_mut<V: key>(table: &mut TypeTable): &mut V;
+
+    #[private_generics(T)]
+    public fun remove<V: key>(table: &mut TypeTable): V;
+
+    #[private_generics(T)]
+    public fun contains<V: key>(table: &TypeTable): bool;
+}
+```
+
+The methods in TypeTable rely on the type safety guarantee of `private_generics`, with constraints similar to Move's global storage instructions.
+
+### AccountStorage
+
+AccountStorage is an abstraction of a user's storage space in Move, which contains two Tables, the Resource Table and Module Table, allowing Resources and Modules to be directly manipulated in Move without the need for global storage instructions.
+
+```move
+struct AccountStorage has key {
+    resources: TypeTable,
+    modules: Table<String, vector<u8>>,
+}
+```
+
+It mainly provides the following API:
+
+```
+module moveos_std::account_storage{
+
+    #[private_generics(T)]
+    /// Move a resource to the account's storage
+    /// This function equates to `move_to<T>(&signer, resource)` instruction in Move
+    public fun global_move_to<T: key>(ctx: &mut StorageContext, account: &signer, resource: T);
```

Comment:
I fee like there are 2 sets of object operation methods? for example if I wanna create a resource under my account, I have 2 options right?
 1. `global_move_to<T: key>(ctx: &mut StorageContext, account: &signer, resource: T)` 
 2. `add<T: key>(this: &mut ObjectStorage, obj: Object<T>)` 

As a dapp dev, will I be able to use both of them? if so, which one should I use under what circumstances.

Reply:
ObjectStorage is for ObjectID -> Object store, and AccountStorage is for StructType -> Struct store.